### PR TITLE
Fix exception when using MSC3030 to look for remote federated events before room creation

### PR DESCRIPTION
**Pull request by MadLittleMods**
Fix exception when using MSC3030 to look for remote federated events before room creation.

Complement tests: https://github.com/matrix-org/complement/pull/405

This happens when you have some messages imported before the room is created.
Then use MSC3030 to look backwards before the room creation from a remote
federated server. The server won't find anything locally, but will ask over
federation which will have the remote event. The previous logic would
choke on not having the local event assigned.

```
Failed to fetch /timestamp_to_event from hs2 because of exception(UnboundLocalError) local variable 'local_event' referenced before assignment args=("local variable 'local_event' referenced before assignment",)
```

### Dev notes

```
synapse_main | 2022-07-06 13:32:48,807 - synapse.handlers.room - 1369 - INFO - GET-6 - get_event_for_timestamp: locally, we found event_id=None closest to timestamp=1657114367164
synapse_main | 2022-07-06 13:32:48,808 - synapse.handlers.room - 1404 - INFO - GET-6 - get_event_for_timestamp: locally, we found event_id=None closest to timestamp=1657114367164 which is next to a gap in event history so we're asking other homeservers first

synapse_main | 2022-07-06 13:32:48,900 - synapse.handlers.room - 1425 - INFO - GET-6 - get_event_for_timestamp: response from domain(hs1)=TimestampToEventResponse(event_id='$6N0qwxOfLXTf7BZRzc5J9LXEPz1dXbWhD1IeAhecNRM', origin_server_ts=1641189600000, data={'event_id': '$6N0qwxOfLXTf7BZRzc5J9LXEPz1dXbWhD1IeAhecNRM', 'origin_server_ts': 1641189600000})
synapse_main | 2022-07-06 13:32:48,903 - synapse.handlers.room - 1464 - WARNING - GET-6 - Failed to fetch /timestamp_to_event from hs1 because of exception(UnboundLocalError) local variable 'local_event' referenced before assignment args=("local variable 'local_event' referenced before assignment",)
```

The `got` is because it has those state events locally even though that user joined after those events. We want the imported event which has a timestamp before the creation event.
```
                   -- oldest events --
                   $azd78vChQWi-mL8ujkvYDaGUoBgf5LeiJtcd96SjuVw (1657114367199) - m.room.create
                   $nrXeQWObbNdV1bdT_aV5imUHm-Y7RJjv4t2Bf2Q5t04 (1657114367381) - m.room.member
                   $_8LZSYqibJvowol8p9bkd-tFzUEKrUZ40I6ZdBXLu88 (1657114367431) - m.room.power_levels
                   $ZPCpSJy-24c_WIHTr3RFmVETRE5yzHl9xT27ek5DkZ4 (1657114367479) - m.room.join_rules
             (got) $_zb4erdxtTYrqOvRdqxTKONKzEk8iq9iPDfjx9Ey6VY (1657114367522) - m.room.history_visibility
                   $l8DQs4X8PiZqqsHCDHMnkjmPMJruwtE3biWXW7TJPCc (1657114367572) - m.room.message
                   $gQTxarzOc3rzF1FLr3pcCNyxJtMdwbfnSZraLLXGa8U (1657114367649) - m.room.message
                   $kRFvZBZT9p4nLqSGRfFmTUER-GKosW6aYpL_DdSKbBM (1657114367702) - m.room.member
  (want, imported) $6N0qwxOfLXTf7BZRzc5J9LXEPz1dXbWhD1IeAhecNRM (1641189600000) - m.room.message
                   $8h4X8YiVmOqTj_HSRUCY4LuiO8oA3xOVR0WjGMLPgJU (1657114368559) - m.room.member
                   -- newest events --
```

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/handlers/room.py [hunk 1]
```diff
@@ -1461,7 +1462,7 @@ async def get_event_for_timestamp(
                         ex.args,
                     )
 
-        if not local_event_id:
+        if not local_event_id or not local_event:
```
Comment (squahtx) [inline synapse/handlers/room.py:1465]: I think it's impossible to have a `local_event_id` but no `local_event`?

```suggestion
        if not local_event_id:
```

Comment (squahtx) [reply synapse/handlers/room.py:1465]: That's unfortunate.
We could add an `assert local_event is not None` below?

Comment (MadLittleMods) [reply synapse/handlers/room.py:1465]: > We could add an `assert local_event is not None` below?

Feels like that's not better though.

I'd rather throw that descriptive error if `local_event` is ever `None` and have the single logic branch.

Comment (MadLittleMods) [reply synapse/handlers/room.py:1468]: Added 👍

https://github.com/matrix-org/synapse/blob/f845f07123e1fe09449fabad467ff4c3159dbaf9/synapse/handlers/room.py#L1465-L1468

Comment (squahtx) [reply synapse/handlers/room.py:1468]: Thanks, looks good!

**Review by squahtx — APPROVED** [review]

File: synapse/handlers/room.py [hunk 2]
```diff
@@ -1375,6 +1375,7 @@ async def get_event_for_timestamp(
         # the timestamp given and the event we were able to find locally
```
Comment (MadLittleMods) [inline synapse/handlers/room.py:1375]: Thanks for the review @squahtx 🐰

**Pull request merged**